dali/pipeline/operator/eager_operator.h

```diff
@@ -165,9 +169,28 @@ class DLL_PUBLIC EagerOperator {
                        ", instance name: \"", name_, "\", encountered:\n", what);
   }
 
-  static inline int GetDefaultNumThreads() {
-    int num_cores = std::thread::hardware_concurrency();
-    return num_cores < 6 ? num_cores : 6;
+  static inline std::shared_ptr<ThreadPool> GetSharedThreadPool() {
+    std::lock_guard<std::mutex> lock(shared_thread_pool_mutex_);
+
+    if (!shared_thread_pool_) {
+      int num_cores = std::thread::hardware_concurrency();
+      num_cores = num_cores < 6 ? num_cores : 6;
+
+      shared_thread_pool_ =
+          std::make_shared<ThreadPool>(num_cores, CPU_ONLY_DEVICE_ID, false, "EagerOperator");
+    }
+
+    return shared_thread_pool_;
+  }
+
+  static inline std::shared_ptr<CUDAStreamLease> GetSharedCudaStream() {
+    std::lock_guard<std::mutex> lock(shared_cuda_stream_mutex_);
```

Comment:
nitpick: You could use a shared lock here (as opposed to unique_lock), so that several threads can access the getter at the same time, and only the updater function will use an exclusive lock.

Reply:
We cannot use shared lock here (at least not just by replacing the current lock) because that way we would initialize thread pool many times.